### PR TITLE
Feature: Custom model timeline dates

### PR DESCRIPTION
**Pull request by samoddball**
# MINT-3680

## Description

- feature branch for custom date portion of the existing timeline
- includes UI, BE, DB migrations, and email notifications
- figma: https://www.figma.com/design/qXhO12xSSkDF4LaToGKJoV/MINT-Blue-Sky-Ideas?node-id=6818-21030&p=f&t=r6bePDwEWPWdceGO-0

## How to test this change

1.  ensure all build/lint/tests pass
2. checkout branch and rebuild locally, including db seed
3. search model plans locally for "timeline", and only one should return - that is the seeded model plan with custom date mock data
4. nav to the timeline of that model plan, ensure looks like figma
5. try bulk editing a few dates at once, including multiple custom dates and standard dates, confirm successful completion (ensure email notification goes out and looks as expected)
6. try adding a date, ensure matches figma and functions as expected (ensure email notification goes out and looks as expected)
7. try editing a date (via separate screen), ensure matches figma and functions as expected (ensure email notification goes out and looks as expected); Then try edit a date on timeline page without clicking 'edit' button, then click return, the changes should also be saved as well.
8. try removing a date - ensure modal matches figma and functions as expected

## PR Author Checklist

- [ ] I have provided a detailed description of the changes in this PR.
- [ ] I have provided clear instructions on how to test the changes in this PR.
- [ ] I have updated tests or written new tests as appropriate in this PR.
- [ ] Updated the [BRUNO Collection](../query_examples/MINT) if necessary.

## PR Reviewer Guidelines

- It's best to pull the branch locally and test it, rather than just looking at the code online!
- When approving a PR, provide a reason _why_ you're approving it
  - e.g. "Approving because I tested it locally and all functionality works as expected"
  - e.g. "Approving because the change is simple and matches the Figma design"
- Don't be afraid to leave comments or ask questions, especially if you don't understand why something was done! (This is often a great time to suggest code comments or documentation updates)
- Check that all code is adequately covered by tests - if it isn't feel free to suggest the addition of tests.

### CODE REVIEW

**Comment by StevenWadeOddball** [comment]
@samoddball and @yiranli624 can you fill out the PR instructions please?

File: migrations/V268__Add_Custom_Timeline_Date.sql [hunk 1]
```diff
@@ -0,0 +1,31 @@
+CREATE TYPE CUSTOM_TIMELINE_DATE_TYPE AS ENUM (
+    'SINGLE',
+    'RANGE'
+);
+
+CREATE TABLE IF NOT EXISTS custom_timeline_dates (
```
Comment (StevenWadeOddball) [inline migrations/V268__Add_Custom_Timeline_Date.sql]: NIT: the MINT database prefers non plural table names for one to many relations (EG where each row represents the type). You will see records like plan basics with a plural name, but that is in refernce to a row that holds multiple `basics` entries. Could you update please

File: migrations/V269__Add_Custom_dates_note.sql
Comment (StevenWadeOddball) [inline migrations/V269__Add_Custom_dates_note.sql:1]: This might have been co-pilot feedback to add this column in a separate migration. It's unnecessary for a feature branch. Could you please combine with the primary migration?

File: pkg/storage/loaders/custom_timeline_date_loader.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+	data, err := storage.CustomTimelineDateGetByIDLoader(loaders.DataReader.Store, ids)
+	if err != nil {
+		return errorPerEachKey[uuid.UUID, *models.CustomTimelineDate](ids, err)
+	}
+
+	getKeyFunc := func(data *models.CustomTimelineDate) uuid.UUID {
+		return data.ID
+	}
+
+	return oneToOneDataLoader(ids, data, getKeyFunc)
```
Comment (StevenWadeOddball) [inline pkg/storage/loaders/custom_timeline_date_loader.go:60]: Since this is for a deletable entity, I wonder if it should use a variant that allow nil returns? There is a variant in the codebase called oneToOneDataLoaderAllowNil, but it looks like it's commented out because it's not in use. 

I think this might help clarify some behavior when getting meta data for a change history record. I think it could still work the way you have it though as well

**Review by StevenWadeOddball — COMMENTED** [review]
@samoddball , I made just a few suggestions on structure from what I saw in another branch.

File: migrations/V268__Add_Custom_Timeline_Date.sql (hunk 1 above)
Comment (samoddball) [reply migrations/V268__Add_Custom_Timeline_Date.sql]: updating in https://github.com/CMS-Enterprise/mint-app/pull/2423 - thanks!

File: migrations/V269__Add_Custom_dates_note.sql
Comment (samoddball) [reply migrations/V269__Add_Custom_dates_note.sql:1]: updated in https://github.com/CMS-Enterprise/mint-app/pull/2423

File: pkg/storage/loaders/custom_timeline_date_loader.go (hunk 2 above)
Comment (samoddball) [reply pkg/storage/loaders/custom_timeline_date_loader.go:60]: discussed this - this could be a paradigm shift, so i left alone in this patch

File: src/features/ModelPlan/Timeline/index.tsx [hunk 3]
```diff
@@ -549,44 +577,66 @@ const Timeline = () => {
...
+                        </div>
+
+                        <Button
+                          type="button"
+                          className="usa-button usa-button--unstyled"
+                          onClick={() => {
+                            if (formikRef.current) {
+                              formikRef.current.values =
+                                formikRef.current.initialValues;
+                            }
```
[the remark this answers is not in the document]
Comment (yiranli624) [reply src/features/ModelPlan/Timeline/index.tsx:627]: nope, reset doesnt work in this case.

File: pkg/sqlqueries/SQL/custom_timeline_date/update_dates_by_ids.sql [hunk 4]
```diff
@@ -0,0 +1,25 @@
+UPDATE custom_timeline_dates AS ctd
+SET
+    start_date = COALESCE(updates.start_date, ctd.start_date),
+    end_date = updates.end_date,
+    modified_by = :modified_by,
+    modified_dts = CURRENT_TIMESTAMP
```
[the remark this answers is not in the document]
Comment (samoddball) [reply pkg/sqlqueries/SQL/custom_timeline_date/update_dates_by_ids.sql:6]: we include both dates on a range update, even if only one was changed,  so this wouldn't be a problem

**Review by Peyton232 — APPROVED** [review]
migration V268 next in sequence with no collisions.
builds cleanly locally and tests pass
Code looks good from a quick pass 

Looks good to merge from my end
